dtm0/doc/dtx.txt

```diff
@@ -0,0 +1,409 @@
+}
+
+ast_cb dix_cas_dtx_executed(...) {
+	struct cas_fop fop;
+
+	if (dixreq->dtx.state == InProgress) {
+		dtx_set_state(dixreq->dr_dtx, Executed);
+		/* m0 client op uses the same sm group */
+		/* set_state() causes a clink to be called, and it causes
+		   asts to be called where M0 client calls op_executed(),
+		   and changes the op state to Executed.
+		*/
+	}
+	
+	/* Thread-safety: we may need to lock() d_txr
+	if it might be modified by DTM0 service. */
+	dixreq->dr_dtx->d_txr.t_pa[fop.reply.source].pa_executed = true;
+}
+
+dix_cas_rop_clink_cb(...) {
+	struct dix_req *dix_req = ...;
+	dixreq->dr_nr_completed++;
+	
+	if (dixreq->dtx != NULL) {
+		post ast dix_cas_dtx_executed(...);
+	}
+
+	/* preserve the existing mechanism */
+	if (nr_completed == nr_requests) {
+		post ast dix_rop_completed(...);
+	}
+}
+
+struct dtm_service {
+	struct dtm_log   ds_log;
+	struct sm_group *ds_sm_group;
+};
+
+ast_cb client_dtx_stable(op,...) {
+	struct m0_op *op;
+	
+	op->op_stable();
+	op->state = Stable;
+}
+
+clink_cb m0_client_dtx_clink_cb(dtx, ...) {
+	post ast dtx->sm_group client_dtx_stable(dtx->op);
+}
+
+ast_cb ast_on_dtx_stable(dtx, ...) {
+	dtx_set_state(dtx, Stable);
+	/* Assume: a dedicated clink is subscribed to dtx->sm. */
+	// causes on_stable_clink_cb to be called
+}
+
+ast_cb process_persistent_notice(fop, ...) {
+	struct dtm_service service = fop->service;
+
+	// Take a read lock on the container of records.
+	rw_read_lock(service->ds_log);
+
+	record = service->ds_log->find(fop->data);
+
+	// Take a write lock on the log entry.
+	rw_write_lock(record);
+
+	record->txr[fop->source.id].persistent = true;
+	
+	// is_stable = forall(i=0..N-1, record.txr[i].persistent == true)
+	if (!is_stable(record)) {
+		goto out;
+	}
+
+	dtx_sm_group = record->dtx->d_sm_group;
+
+	// We cannot directly modify m0 client data
+	// because we are running within DTM service SM group.
+	// We need a bridge between DTM service SM group and
+	// M0 Client SM group. Moreover, we need to change
+	// the state of DTX first, and then change the state
+	// of M0 client operation (and also we should call op_stable).
+
+	// Ast(DTM Service) -> Ast(DTX): a bridge between SM groups
+	// Ast(DTX) -> Clink(Client) -> Ast(Client): a bridge between
+        //                                           the components 
+	//                                           (Client and DTX).
+	// It gives us the following sequence:
+	//	Ast1 -> Ast2 -> Clink -> Ast3.
+	
+	ast post dtx_sm_group ast_on_dtx_stable(...);
+
+out:
+	unlock(record);
+	unlock(service->ds_log);
+}
+
+// The function is called by RPC when it gets a PERSISTENT notice
+// from another node.
+fop_process_persistent_notice(...) {
+	post ast dtm_service.sm_group process_persistent_notice(fop);
+}
+
+
+--------------------------------------------------------------------------------
+# NOTES
+
+CS should check the previous timestamp, and it should wait until the value
+changes. It could be a busy loop.
+
```

Comment:
I propose to add one more note:

We need to preserve existing logic without DTM by adding of conditions like
        if (op->op_dtx)
                ......
by the following reasons:
 - we have a read requests that go without DTX but use common code;
 - we have a meta requests that possibly will not use DTX;
 - we need to support backward compatibility for easier testing and other purposes.

Reply:
Added & updated